### DISCUSSION TITLE
Button/WS Limits?

### DISCUSSION THREAD

**Discussion by CommanderB** [ybizeul/StreamDeckWS#17] [Q&A]
I am butting up against some odd errors which I have been going round in circles on. This is a simple example of one of the issues - 

I can write to a button when using hard coded context, but errors when using the `sd output` node (which works for all other buttons).

This works

![Untitled 2](https://user-images.githubusercontent.com/18680188/105612489-8305d680-5d8a-11eb-9fda-8849cade34e8.png)

And msg.payload sent back to the WS is

![Untitled](https://user-images.githubusercontent.com/18680188/105612499-931db600-5d8a-11eb-9276-c9a8bcb895af.png)

When I change this to use the `sd output` node configured as follows

![Untitled 4](https://user-images.githubusercontent.com/18680188/105612516-b2b4de80-5d8a-11eb-9a77-0382aca9c6bf.png)

The node errors (and the debug node doesnt fire) as below

![Untitled 3](https://user-images.githubusercontent.com/18680188/105612527-cbbd8f80-5d8a-11eb-9215-5134b1e54ff3.png)

I have similar problems when trying to write an image to the button. 

I also seem to have problems when I configure a 5th WS node, everything stops working. Removing the 5th WS restores the 4 working button/WS flows to work on the SD. I will try to demonstrate that in a different message

**Comment by ybizeul** [comment]
Can you check the content of Context Data for variable `streamDeckContexts` ?

![image](https://user-images.githubusercontent.com/182035/105613672-46ef6780-5dc4-11eb-85b8-a4ee439091ef.png)

if `btnTime` is unknown it's probably because no `sd in` node ever saw that node `willAppear`, `keyUp`, etc. before

**Comment by CommanderB** [comment]
Yup, its not there, nor is btnTemp or, another one I am hardcoding around - 

![Untitled 7](https://user-images.githubusercontent.com/18680188/105614324-a9317380-5d96-11eb-9ecc-33de1dc3ac62.png)
![Untitled 11](https://user-images.githubusercontent.com/18680188/105614508-dd596400-5d97-11eb-8171-8ecd7e053bd7.png)

This is the flow that captures button presses, btnLight works. (Key is a subflow with an `sd input` and switch on key down/up/willAppear which is used in several places)

![Untitled 8](https://user-images.githubusercontent.com/18680188/105614424-6e7c0b00-5d97-11eb-883b-fc2f8d9d2908.png)

And pressing btnTime or btnTemp works.

![Untitled 9](https://user-images.githubusercontent.com/18680188/105614448-8d7a9d00-5d97-11eb-9294-c9cc4c3471d7.png)
![Untitled 10](https://user-images.githubusercontent.com/18680188/105614450-923f5100-5d97-11eb-9020-747b945e69c3.png)

**Comment by CommanderB** [comment]
So its the subflow!!! My bad. The scope of `streamDeckContexts` isnt making it out of the subflow - 

Changing to this

![Untitled 12](https://user-images.githubusercontent.com/18680188/105614731-1cd48000-5d99-11eb-9595-5bcbb5d5ef53.png)

Has it working - 

![Untitled 13](https://user-images.githubusercontent.com/18680188/105614751-45f51080-5d99-11eb-94e5-209c486f9e6a.png)

Sorry. Thats me not thinking/fully understanding scope in Node-Red

**Reply by ybizeul** [reply]
States should actually be managed by a configuration node, node in the flow, that would be the proper way to do it, I didn't know better at the time, I'll work on a fix
